### DISCUSSION TITLE
Solve problem on a lower dimensional subdomain

### DISCUSSION THREAD

**Discussion by HelgeGehring** [kinnala/scikit-fem#982] [Q&A]
I have a 2D (and maybe somewhen 3D) mesh and I'd like to solve a problem on a 1D (2D) boundary of the mesh.

For example if I have the mesh
![image](https://user-images.githubusercontent.com/42973196/202542772-74a6c9f0-0b49-496c-b881-5abc42d6fde4.png)

and define a variable on that which looks like
![image](https://user-images.githubusercontent.com/42973196/202545325-843c4679-cf60-481a-951c-85fb5dd37fd2.png)

I'd like to interpolate the variable to that dark blue line in the center, solve the lower dimensional problem (i.e. finding the eigenvectors at that facet in the lower dimension) and use the solution I found in `condense` to eliminate those degrees of freedom from the 2D (3D) problem and thus define the solution I'll find.

At the moment I'm thinking about something like this:
```python
from skfem import *

mesh = MeshTri().refined(3) \
    .with_boundaries({'center': lambda x: x[1] == 0.5}, boundaries_only=False) \
    .with_subdomains({'middle': lambda x: (abs(x[0] - .5) < .25) * (abs(x[1] - .5) < .25)})

mesh.draw(boundaries=True).show()

basis = Basis(mesh, ElementTriP0())
epsilon = basis.zeros() + 3
epsilon[basis.get_dofs(elements='middle')] = 12
basis.plot(epsilon, colorbar=True).show()

basis_center = FacetBasis(mesh, ElementTriP0(), facets='center')
epsilon_center = basis_center.project(basis.interpolate(epsilon))
```
But it doesn't wok like that and I don't really know yet how to approach this better.

Ideally, I'd like to solve a 1D problem like
![image](https://user-images.githubusercontent.com/42973196/202545486-4dfb8f11-1928-49b4-aa2b-3736f49601b4.png)
on a boundary and use this to define the solution of the full problem.

Could you give me some advise how to tackle this problem?

**Comment by kinnala** [comment]
I think your idea of using FacetBasis on 'center' can work. However, anything you assemble with 'basis_center' will have as many rows/columns as there are DOFs in the full mesh. In order to just solve/project on the interface, you need to look up those DOF numbers on 'center' using get_dofs and slice the resulting matrices. I think project has this additional kwarg 'facets' which can be set to 'centrer'.

**Reply by HelgeGehring** [reply]
Thanks, I'll try :)

The ElementTriP0 doesn't have dofs at the facets, that would create a problem, right? Would I need to Interpolate the epsilon first on a P1 element? Or is there a way to do it directly?

**Reply by kinnala** [reply]
As epsilon is just data, you don't need to project it I think. You can interpolate using basis_center directly.

**Reply by HelgeGehring** [reply]
Okay, thanks! That worked for me, I get the epsilon on that line.

I can calculate an eigenvector on that line with a ElementTriP1 like I would to with an ElementLineP1 (I get the expected eigenvalues), but I had to change `inner(grad(u), grad(v))` to `inner(grad(u)[0], grad(v)[0])` to make it work. For a horizontal line that's fine, but how could I generalize that? Can I somehow get a vector in direction of the facet?

For the larger problem I use as a basis `Basis(mesh, ElementTriN1() * ElementTriP1())`, while I use for the subproblem a Basis `basis_source_1d` with the Element `ElementTriP1()`, as the calculation only needs to have one component. This component I'd like to set afterwards as the y-component of a vector of the Basis for the complete Problem to set it as `x` for `solve`. For that I use:
```python
x0 = basis_source.project((np.array((basis_source_1d.interpolate(0*xs[:,-1]), basis_source_1d.interpolate(xs[:,-1]))), basis_source_1d.interpolate(0*xs[:,-1])))
```
where  basis_source is defined as `FacetBasis(mesh, ElementTriN1() * ElementTriP1())` and xs are my eigenvectors.
which I use in condense like 
```python
condense(..., D=basis_source.get_dofs(facets='source'), x=x0)
```
but it doesn't look like I would expect it. If I calculate with the ElementLineP1 it looks like a Gaussian, but here the y-component looks like
![image](https://user-images.githubusercontent.com/42973196/202647605-7a0e2f95-5299-47ba-8f9f-32e47e09dacc.png)
and the x-component doesn't seem to be zero

Would I need to do something different here? I thought that I maybe need `ElementTriN2`  istead of `N1` to represent the P1 better? It also seems to restrict all other Dofs to 0 in the Components represented by ElementTriN2, did I maybe choose `D` wrong?
(You can find the complete script here: https://github.com/HelgeGehring/femwell/blob/main/femwell/fefd.py)

**Reply by kinnala** [reply]
Normal vector is available as `w.n` when assembling with `FacetBasis` so if you want the tangential vector you can do something like `t = np.array([w.n[1], -w.n[0]])` and then calculate `dot(grad(u), t)`. However, you probably want to make sure that `'center'` is then oriented so that the normal vector is always pointing to the correct direction.

You can construct oriented boundaries like this:
```python
In [7]: bnd1 = m.facets_satisfying(lambda x: x[1] == 0.5, normal=[0, 1])

In [8]: bnd1
Out[8]: OrientedBoundary([17, 21, 22, 26, 88, 89, 94, 95])

In [9]: M = m.with_boundaries({'center': bnd1})

In [10]: M.boundaries
Out[10]: 
{'left': array([ 1,  5, 14, 15, 36, 37, 54, 55]),
 'bottom': array([ 0,  2, 10, 11, 32, 33, 40, 41]),
 'right': array([ 4,  8, 24, 25, 50, 51, 68, 69]),
 'top': array([ 7,  9, 28, 29, 64, 65, 72, 73]),
 'center': OrientedBoundary([17, 21, 22, 26, 88, 89, 94, 95])}
```

**Reply by kinnala** [reply]
Sometimes those mesh formats might come with oriented subsets of facets, but this is not an issue we tried to tackle here because there are so many special cases and it's more "geometry problem" than "finite element problem".

**Reply by HelgeGehring** [reply]
Nice, the tangent vector works great! Might it make sense to also add something like w.t?
And I've just checked, gmsh gives me OrientedBoundaries :)

**Reply by kinnala** [reply]
About your second question: If I understand correctly `basis_source_1d` is `FacetBasis` and `basis_source` is `Basis`. Then `interpolate`/`project` are not compatible between those two different basis types. What `interpolate` does is interpolating the solution at the quadrature points. Since `Basis` and `FacetBasis` have completely different quadrature points (first is 2D and second is 1D), it's simply not going to work this way.

We need to carefully think how you want to do the extension from 1D to 2D and then come up with a solution.

**Reply by kinnala** [reply]
I've been thinking about `w.t`. However, since in 3D you would need `w.t1` and `w.t2` and some kind of automatic decision on what is the direction of `w.t1`, I've got stuck there.

**Reply by HelgeGehring** [reply]
basis_source is here also a FacetBasis (everything with basis_source in the name is a FacetBasis), which I'd like to use to define the Dofs at the facet, which I extract from the larger Mesh.

My thinking was:
- for the complete problem I use ElementN1*ElementP1, so if I want to set Dofs in the complete problem `basis_source` needs the same element.
- As I only need to solve one component to define the mode (TE/TM modes separate for 2D problems, which leads for TE modes to only one non-zero component in the E-field), I use a reduced FacetBasis with ElementP1 to solve for the mode on this Facet
- I need to project it back on the FacetBasis which has the same element as the Basis for the complete problem
- I set the Dofs in `solve`

(for physical correctness: it's not really a source, that would have been on the rhs of the implemented equation. However, setting the mode somewhere has a quite similar effect, I'll rename it later on)

**Reply by HelgeGehring** [reply]
I also don't see a general way to decide the direction for `w.t1`. But I'd guess an alternative at least for these problems would be to just use
```python
inner(grad(u),grad(v))-inner(dot(grad(u),w.n),dot(grad(v),w.n))
```

**Reply by HelgeGehring** [reply]
Okay, I think I start getting it work :)
But at the moment I can set two of the components:
- I can get the DOFs represented by the ElementTri1, as the DOFs associated with the quadratures at the facet are also on the facet
- The same for the DOFs along the Facet which are represented by ElementTriN1. As the quadrature points on the facet define the dofs for the values along the facet.

But there are some problems/questions:
- I don't know how to get the DOFs associated with the normal component on the facet. As it's represented by ElementTriN1 the DOFs are not on the facet, but close. Is there something like `get_dofs_defining(facet='source')`? If I look at the result of the projection I see that there are non-zero values.
- How do I get DOFs for only one component of a Basis with ElementComponent (ElementVector)? with the current way I get all the DOFs associated with the facet/... if I use `get_dofs`
- Even through not clean, but maybe efficient for testing: Is there some way how I get the DOFs for all non-zero values of a projected vector?

**Reply by kinnala** [reply]
> * I don't know how to get the DOFs associated with the normal component on the facet. As it's represented by ElementTriN1 the DOFs are not on the facet, but close. Is there something like `get_dofs_defining(facet='source')`? If I look at the result of the projection I see that there are non-zero values.

I don't understand what you mean. For ElementTriN1 there are no DOFs associated with the normal component.

> * How do I get DOFs for only one component of a Basis with ElementComponent (ElementVector)? with the current way I get all the DOFs associated with the facet/... if I use `get_dofs`

Try to look at the object returned by get_dofs. It allows further filtering the results.

> * Even through not clean, but maybe efficient for testing: Is there some way how I get the DOFs for all non-zero values of a projected vector?

np.nonzero?

**Reply by HelgeGehring** [reply]
If I look at https://defelement.com/elements/examples/triangle-Nedelec-1.html the tangential component of the lower facet is defined at the lower edge - the normal component is defined by the other two facets, right? or do I understand something wrong?

I'll have a look, thanks!

np.nonzero sounds good! I started with np.where, but was confused that .flatten is called afterwards. np.array(np.nonzero(...)) works, thanks!

**Reply by kinnala** [reply]
Alright, when you say "DOF associated with" I'm automatically thinking how DOFs are associated with the topological entities of the mesh. There isn't anything like what you are asking but if you want the other facets given a set of boundary facets you can use Mesh.f2t and Mesh.t2f to find those. (f2t = facet2triangle, t2f = triangle2facet)
